src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,473 @@
+        {
+            _undoHistoryRegistry = undoHistoryRegistry;
+            _editorOperationsFactoryService = editorOperationsFactoryService;
+        }
+
+        public string DisplayName => CSharpEditorResources.Complete_statement;
+
+        public void ExecuteCommand(TypeCharCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            if (args.TypedChar != ';')
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            var caret = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caret.HasValue)
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            var document = caret.Value.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            var root = document.GetSyntaxRootSynchronously(executionContext.OperationContext.UserCancellationToken);
+            var caretPosition = caret.Value.Position;
+            
+
+            var token = GetToken(root, caretPosition, caret);
+
+            var currentNode = token.Parent;
+            // if cursor is right before a closing delimiter, make sure you start with node outside of delimiters
+            if ((token.Kind() == SyntaxKind.OpenBraceToken || token.Kind() == SyntaxKind.OpenBracketToken || token.Kind() == SyntaxKind.OpenParenToken) && token.Span.End == caret.Value.Position)
+            {
+                currentNode = currentNode.Parent;
+            }
+
+            if (currentNode == null)
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            if (GetEnclosingArgumentList(currentNode, token, caret, caretPosition, syntaxFacts) == null)
+            {
+                nextCommandHandler();
+                return;
+            }
+
+            var lastDelimiterPosition = -1;
+            var finalDelimiterNeedsSemicolon = false;
+
+            // work your way out, verifying all delimeters exist until you reach statement syntax that requires a semicolon
+            while (!ReachedSemicolonSyntax(currentNode, syntaxFacts, ref finalDelimiterNeedsSemicolon))
+            {
+                if (!ClosingDelimiterExistsIfNeeded(currentNode, ref lastDelimiterPosition))
+                {
+                    nextCommandHandler();
+                    return;
+                }
+
+                if (currentNode.Parent == null)
+                {
+                    nextCommandHandler();
+                    return;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            if (currentNode.Ancestors().Any(n => n.Kind() == SyntaxKind.ForStatement))
```

Comment:
❕ This will prevent the feature from working inside any `for` statement, including things like the following:

```csharp
for (int x = 0; x < 10; x++)
{
    x.ToString($$)
}
```